### PR TITLE
feat: add module `evidence`

### PR DESCRIPTION
**Pull request by dmvict**
The module has a few main features/disadvantages: 
- used `Any` in genesis and generic for transforming it to concrete type 
- runtime validation of genesis structs 
- router that applies generics related to context `DB, SK` and propagates the generics on `Keeper` and `ABCIHandler` structs. We can't skip it because we don't know these types. And here I've added struct `Handler` with quite ugly api. I was trying to hide the concrete context types and do not extend number of generics in the struct 
- used `Rc` for handler routines. We may replace it to `Arc` if it is required.

### CODE REVIEW

File: x/evidence/src/keeper.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            self.router = Some(router);
+            Ok(())
+        }
+    }
+
+    /// genesis initializes the evidence module's state from a provided genesis
+    /// state.
+    pub fn init_genesis(&self, ctx: &mut InitContext<'_, DB, SK>, genesis: GenesisState) {
+        if let Err(e) = genesis.validate::<E>() {
+            panic!("failed to validate evidence genesis state: {e}");
```
Comment (NYBACHOK) [inline x/evidence/src/keeper.rs:88]: What do you think? Is it worth to return errors from keeper and panic only in abci handler or it more annoying?

File: x/evidence/src/types/mod.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+mod query;
+mod router;
+mod tx;
+pub use router::*;
+
+//
+
+pub trait Evidence: Message {
+    fn route(&self) -> String;
+    fn r#type(&self) -> String;
```
Comment (NYBACHOK) [inline x/evidence/src/types/mod.rs:13]: kind?

File: x/evidence/src/types/router.rs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    Block(Rc<HandlerFn<BlockContext<'a, DB, SK>, E>>),
+}
+
+impl<'a, DB, SK, E> Handler<'a, DB, SK, E>
+where
+    E: Evidence,
+    SK: StoreKey,
+    DB: Database,
+{
+    pub fn new_init(handler_fn: HandlerFn<InitContext<'a, DB, SK>, E>) -> Self {
```
Comment (NYBACHOK) [inline x/evidence/src/types/router.rs:50]: Hmmm. This all structure is... okay. What about implementing `From`?

**Comment by NYBACHOK** [comment]
This code is not common for us, but hard to say without all context why you chose it. Maybe you could describe which issues you have with generics and etc?

File: x/evidence/src/types/mod.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    fn height(&self) -> i64;
+}
+
+pub trait EvidenceHandler<E: Evidence>
+where
+    <E as std::convert::TryFrom<Any>>::Error: std::fmt::Debug,
+{
+    type Error;
+
+    fn handle<CTX: TransactionalContext<DB, SK>, DB: Database, SK: StoreKey>(
```
Comment (joneskm) [inline x/evidence/src/types/mod.rs:35]: Do you prefer to have two traits or could the `handle` be part of the `Evidence` trait?

Comment (dmvict) [reply x/evidence/src/types/mod.rs:25]: Yeah. It's good idea. I've merged the traits.

File: x/evidence/src/types/mod.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+where
+    <Self as std::convert::TryFrom<Any>>::Error: std::fmt::Debug,
+{
+    // TODO: uncomment or remove
+    // fn route(&self) -> String;
+    // Original method is named `type`, replaced as inner interface
+    fn kind(&self) -> String;
+    fn string(&self) -> String;
+    fn hash(&self) -> Hash;
+    fn validate_basic(&self) -> anyhow::Result<()>;
```
[the remark this answers is not in the document]
Comment (dmvict) [reply x/evidence/src/types/mod.rs:22]: In this case I saved the function. We can't force a user to implement validation of his evidence in implementation of `TryFrom<Any>`. 
We can remove the member if we force users to use our evidence type (actually Tendermint evidence type), but then we can skip all generics in the module.

File: x/evidence/src/genesis.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+use gears::core::any::google::Any;
+use serde::{Deserialize, Serialize};
+
+use crate::Evidence;
+
+/// GenesisState defines the evidence module's genesis state.
+#[derive(Debug, Clone, Default, PartialEq, Deserialize, Serialize)]
+pub struct GenesisState {
+    /// evidence defines all the evidence at genesis.
```
[the remark this answers is not in the document]
Comment (dmvict) [reply x/evidence/src/genesis.rs:11]: Added and removed validation.

File: x/evidence/src/types/router.rs (hunk 3 above)
Comment (dmvict) [reply x/evidence/src/types/router.rs]: non actual

File: x/evidence/src/keeper.rs (hunk 1 above)
Comment (dmvict) [reply x/evidence/src/keeper.rs]: Moved

File: x/evidence/src/types/mod.rs (hunk 2 above)
Comment (dmvict) [reply x/evidence/src/types/mod.rs]: Replaced

File: x/evidence/src/types/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (NYBACHOK) [reply x/evidence/src/types/mod.rs:22]: > but then we can skip all generics in the module.
Isn't it good?

File: x/evidence/src/genesis.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+where
+    <E as std::convert::TryFrom<Any>>::Error: std::fmt::Debug,
+{
+    /// evidence defines all the evidence at genesis.
+    pub evidence: Evidences<E>,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, Default)]
+#[serde(try_from = "Vec<Any>")]
+pub struct Evidences<E: Evidence>(Vec<E>)
```
Comment (joneskm) [inline x/evidence/src/genesis.rs:17]: Would it make sense to have the `#[serde(try_from = "Vec<Any>")]` on the Gensis struct and avoid having the `Evidences` struct altogether?

File: x/evidence/src/types/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (joneskm) [reply x/evidence/src/types/mod.rs:22]: I would have thought that most users would be familiar with the idea of doing validation in a `TryFrom` implementation.

**Review by joneskm — APPROVED** [review]

**Review by NYBACHOK — APPROVED** [review]

**Pull request merged**